presto-main/src/main/java/com/facebook/presto/operator/scalar/CharacterStringCasts.java

```diff
@@ -81,24 +87,75 @@ public static Slice charToVarcharCast(@LiteralParameter("x") Long x, @LiteralPar
 
         // if Varchar(x) value length (including spaces) is greater than y, we can just truncate it
-        if (trimmedTextLength + numberOfTrailingSpaces >= y) {
-            return truncateToLength(trimmedSlice, y.intValue());
+        if (codePoints.size() >= y) {
+            // char(y) slice representation doesn't contain trailing spaces
+            codePoints = trimTrailing(codePoints, ' ');
+            List<Integer> codePointsTruncated = codePoints.stream()
+                    .limit(y)
+                    .collect(toImmutableList());
+            return codePointsToSliceUtf8(codePointsTruncated);
         }
-        if (trimmedTextLength == 0) {
-            return EMPTY_SLICE;
+
+        /*
+         * Value length is smaller than same-represented char(y) value because input varchar has length lower than y.
+         * We decrement last character in input (in fact, we decrement last non-zero character) and pad the value with
+         * max code point up to y characters.
+         */
+        codePoints = trimTrailing(codePoints, '\0');
+
+        if (codePoints.isEmpty()) {
+            // No non-zero characters in input and input is shorter than y. Input value is smaller than any char(4) casted back to varchar, so we return the smallest char(4) possible
+            return codePointsToSliceUtf8(nCopies(toIntExact(y), (int) '\0'));
+        }
+
+        codePoints = new ArrayList<>(codePoints);
+        codePoints.set(codePoints.size() - 1, codePoints.get(codePoints.size() - 1) - 1);
+        codePoints.addAll(nCopies(toIntExact(y) - codePoints.size(), Character.MAX_CODE_POINT));
+
+        verify(codePoints.get(codePoints.size() - 1) != ' '); // no trailing spaces to trim
+
+        return codePointsToSliceUtf8(codePoints);
+    }
+
+    private static List<Integer> trimTrailing(List<Integer> codePoints, int codePointToTrim)
+    {
+        int endIndex = codePoints.size();
+        while (endIndex > 0 && codePoints.get(endIndex - 1) == codePointToTrim) {
+            endIndex--;
+        }
+        return ImmutableList.copyOf(codePoints.subList(0, endIndex));
+    }
+
+    private static List<Integer> toCodePoints(Slice slice)
+    {
+        ImmutableList.Builder<Integer> codePoints = ImmutableList.builder();
+        for (int offset = 0; offset < slice.length(); ) {
+            int codePoint = SliceUtf8.getCodePointAt(slice, offset);
+            offset += SliceUtf8.lengthOfCodePoint(slice, offset);
+            codePoints.add(codePoint);
+        }
+        return codePoints.build();
+    }
+
+    private static Slice codePointsToSliceUtf8(List<Integer> codePoints)
+    {
+        int length = codePoints.stream()
+                .mapToInt(SliceUtf8::lengthOfCodePoint)
+                .sum();
+
+        byte[] result = new byte[length];
+        int offset = 0;
+        for (int codePoint : codePoints) {
+            byte[] codePointUtf8 = SliceUtf8.codePointToUtf8(codePoint).getBytes();
```

Comment:
Why not use `lengthOfCodePoint` and `setCodePointAt`, instead? That way, you'd avoid the intermediate slice, byte[] and array copy.